File: wrapper/src/main/java/jack/wrapper/base/mvvm/view/fragment/BaseSimpleFragment.java
package jack.wrapper.base.mvvm.view.fragment;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;

import com.kingja.loadsir.callback.Callback;
import com.kingja.loadsir.core.LoadService;
import com.kingja.loadsir.core.LoadSir;

import cn.jack.library_common_business.loadsir.ViewStateLayout;
import cn.jack.library_common_business.loadsir.callback.CustomCallback;
import cn.jack.library_common_business.loadsir.callback.EmptyCallback;
import cn.jack.library_common_business.loadsir.callback.FailedCallback;
import cn.jack.library_common_business.loadsir.callback.LoadingCallback;
import cn.jack.library_common_business.loadsir.callback.TimeoutCallback;
import jack.wrapper.base.mvvm.view.interf.ILoadSirLisenter;

/**
 * @创建者 Jack
 * @创建时间 2021/3/8 15:03
 * @描述
 */
public abstract class BaseSimpleFragment<V extends ViewDataBinding> extends BaseTopFragment implements ILoadSirLisenter {
    protected V mBinding;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        //DataBindingUtil类需要在project的build中配置 dataBinding {enabled true }, 同步后会自动关联android.databinding包
        mBinding = DataBindingUtil.inflate(inflater, setLayoutRes(), container, false);

        if(isRegisterLoadSir()) {                   //目标：子类自由控制，让整个布局还是指定的view实现状态布局。
            if(isViewRegisterLoadSir()){
                return mBinding.getRoot();
            }else {
                setLoadService(mBinding.getRoot());
                //整个布局页面都会使用状态布局
                return mLoadService.getLoadLayout();
            }
        }else {
            return mBinding.getRoot();
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init();
    }

    private Handler mHandler   = new Handler(Looper.getMainLooper());
    private final int     DELAY_TIME = 1000;

    //由于BaseSimpleFragment没有用到mvvm架构，故需要在view层直接调用该方法
    protected void setViewStateChangeLisenter(ViewStateLayout stateLayout){
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                viewStateChange(stateLayout);
            }
        },DELAY_TIME);
    }

    private void viewStateChange(ViewStateLayout stateLayout){
        if(mLoadService == null){
            return;
        }

        switch (stateLayout){
            case LOADING:
                mLoadService.showCallback(LoadingCallback.class);
                break;
            case SUCCESS:
                mLoadService.showSuccess();
                break;
            case FAILED:
                mLoadService.showCallback(FailedCallback.class);
                break;
            case NET_ERROR:
                //                break;    //暂时
            case TIME_OUT:
                mLoadService.showCallback(TimeoutCallback.class);
                break;
            case CUSTOM:
                mLoadService.showCallback(CustomCallback.class);
                break;
            case EMPTY:
                mLoadService.showCallback(EmptyCallback.class);
                break;
        }
    }

    /**
     * 模板设计模式
     */
    protected final void init() {
        //初始化参数
        prepareParam();
        //初始化数据
        prepareData();
        //设置监听
        prepareListener();
    }

    protected abstract int setLayoutRes();

    private LoadService mLoadService;

    protected void setLoadService(View rootView){
        if(isRegisterLoadSir()) {
            mLoadService = LoadSir.getDefault().register(rootView, new Callback.OnReloadListener() {
                @Override
                public void onReload(View v) {
                    dataReload();
                }
            });
        }
    }

}

File: lib_base/src/main/java/com/jack/lib_base/base/view/BaseSimpleFragment.kt
package com.jack.lib_base.base.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.ViewDataBinding
import cn.jack.library_common_business.loadsir.callback.*
import com.alibaba.android.arouter.launcher.ARouter
import com.jack.lib_base.interfac.ILoadSirLisenter
import com.jack.lib_base.interfac.IStatusSwitchLisenter
import com.jack.lib_base.uistate.LayoutState
import com.jack.lib_base.ext.postCallbackDelayed
import com.jack.lib_base.ext.postSuccessDelayed
import com.jack.lib_wrapper_mvvm.base.view.BaseWrapperFragment
import com.kingja.loadsir.core.LoadService
import com.kingja.loadsir.core.LoadSir

/**
 * @创建者 Jack
 * @创建时间 2022/9/15 20:04
 * @描述
 */
open class BaseSimpleFragment<VB : ViewDataBinding>(override var block: (LayoutInflater) -> VB) :
    BaseWrapperFragment<VB>(block), IStatusSwitchLisenter, ILoadSirLisenter {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (injectARouter()) {
            ARouter.getInstance().inject(this)
        }
    }

    private var mBaseLoadService: LoadService<Any>? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return if (isRegisterLoadSir()) {
            //对整个页面的布局设置状态
            mBaseLoadService = LoadSir.getDefault().register(mBinding.root) {
                dataReload()
            }
            mBaseLoadService!!.loadLayout
        } else {
            super.onCreateView(inflater, container, savedInstanceState)
        }

    }

    /**
     * 对Fragment中指定的View设置状态布局
     */
    protected open fun setTargetLoadService(target: Any?) {
        mBaseLoadService = LoadSir.getDefault().register(target) {
            dataReload()
        }
    }

    //在使用的时候需要谨慎，fragment不要同时 既对整个页面的布局设置状态 又对指定的View设置状态布局
    protected fun setLayoutState(layoutState: LayoutState) {
        when (layoutState) {
            LayoutState.OnLoading -> mBaseLoadService?.postCallbackDelayed(
                LoadingCallback::class.java,
                0
            )
            LayoutState.OnFailed -> mBaseLoadService?.postCallbackDelayed(FailedCallback::class.java)
            LayoutState.OnEmpty -> mBaseLoadService?.postCallbackDelayed(EmptyCallback::class.java)
            LayoutState.OnTimeout -> mBaseLoadService?.postCallbackDelayed(TimeoutCallback::class.java)
            LayoutState.OnCustom -> mBaseLoadService?.postCallbackDelayed(CustomCallback::class.java)
            LayoutState.OnNetError -> mBaseLoadService?.postCallbackDelayed(NetErrorCallback::class.java)
            LayoutState.OnSuccess -> mBaseLoadService?.postSuccessDelayed()
        }
    }

}